### PR TITLE
Feat: select api

### PR DESCRIPTION
**Pull request by Narretz**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
feat

**What is the current behavior? (You can also link to an open issue here)**
select directive behavior is unflexible

**What is the new behavior (if this is a feature change)?**
exposes methods to enable needed flexibility

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by petebacondarwin** [comment]
We should credit the `fix(ngOptions): allow option with ngIf to be removed and re-added  …` commit to the author of the PR.

File: src/ng/directive/select.js [hunk 1]
```diff
@@ -19,10 +19,120 @@ function setOptionSelectedStatus(optionEl, value) {
...
- * added `<option>` elements, perhaps by an `ngRepeat` directive.
+ * The controller for the {@link ng.select select directive}. The controller exposes
+ * a few utility methods that can be used to augment the behavior of a regular or an
+ * {@link ng.ngOptions ngOptions} select element.
+ *
+ * @example
+ * ### Set a custom error when the unknown option is selected
+ *
+ * This example sets a custom error "unknown_value" on the ngModelController
+ * when the select element's unknown option is selected, i.e. when the model is set to value
```
Comment (petebacondarwin) [inline src/ng/directive/select.js]: set to "a" value

**Review by petebacondarwin — APPROVED** [review]
Excellent PR @Narretz

**Comment by Narretz** [comment]
@petebacondarwin after much rebasing, I've attributed the fix to @pbr1111

File: src/ng/directive/select.js [hunk 2]
```diff
@@ -19,10 +19,120 @@ function setOptionSelectedStatus(optionEl, value) {
...
+ *      };
+ *   }])
+ *   .directive('unknownValueError', function() {
+ *     return {
+ *       require: ['ngModel', 'select'],
+ *       link: function(scope, element, attrs, ctrls) {
+ *         var ngModelCtrl = ctrls[0];
+ *         var selectCtrl = ctrls[1];
+ *
+ *         ngModelCtrl.$validators.unknown_value = function(modelValue, viewValue) {
```
Comment (gkalpak) [inline src/ng/directive/select.js]: Nit: Using camelCase is more consistent with other validators.

File: src/ng/directive/select.js [hunk 3]
```diff
@@ -171,6 +281,50 @@ var SelectController =
...
+   * @name select.SelectController#$hasEmptyOption
+   *
+   * @description
+   *
+   * Returns `true` if the select element currently has an empty option
+   * element, i.e. an option that signifies that the select is empty / the selection is null.
+   *
+   */
+  self.$hasEmptyOption = function() {
+    // Presence of the unknown option means it is selected
```
Comment (gkalpak) [inline src/ng/directive/select.js]: This whole comment is incorrect :smiley:

File: test/ng/directive/ngOptionsSpec.js [hunk 4]
```diff
@@ -2536,18 +2536,11 @@ describe('ngOptions', function() {
...
-      expect(element.find('option').length).toBe(2);
-      option = element.find('option').eq(0);
-      expect(option.val()).toBe('');
-      expect(option.text()).toBe('blank');
+      expect(element).toEqualSelect([''], 'object:4');
 
-      scope.$apply(function() {
-        scope.isBlank = false;
-      });
+      scope.$apply('isBlank = false');
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: 👍 ❤️

File: test/ng/directive/selectSpec.js [hunk 5]
```diff
@@ -839,6 +835,111 @@ describe('select', function() {
...
+      // add empty
+      scope.$apply('empty = true');
+      expect(element).toEqualSelect([unknownValue('unmatched')], '', 'x', 'y');
+      expect(selectCtrl.$hasEmptyOption()).toBe(true);
+      expect(selectCtrl.$isEmptyOptionSelected()).toBe(false);
+      expect(selectCtrl.$isUnknownOptionSelected()).toBe(true);
+
+      // unknown -> empty
+      scope.$apply(function() {
+        scope.selected = null;
```
Comment (gkalpak) [inline test/ng/directive/selectSpec.js]: 👎 💔 (for not `$apply('selected = null')` 😛)

**Review by gkalpak — COMMENTED** [review]
Really nice  :100:
(Loved the tests :smiley:)

**Comment by Narretz** [comment]
Thanks for the speedy reviews on this ❤️

**Pull request merged**